solver race

```console
$ go test -race ./solver
...
==================
WARNING: DATA RACE
Read at 0x00c4204c6c60 by goroutine 376:
  runtime.mapiterinit()
      /usr/local/go/src/runtime/hashmap.go:691 +0x0
  github.com/moby/buildkit/solver.(*state).getSessionID()
      /home/suda/gopath/src/github.com/moby/buildkit/solver/jobs.go:67 +0xc4
  github.com/moby/buildkit/solver.(*sharedOp).Exec.func1()
      /home/suda/gopath/src/github.com/moby/buildkit/solver/jobs.go:657 +0x216
  github.com/moby/buildkit/util/flightcontrol.(*call).run()
      /home/suda/gopath/src/github.com/moby/buildkit/util/flightcontrol/flightcontrol.go:95 +0xa8
  github.com/moby/buildkit/util/flightcontrol.(*call).(github.com/moby/buildkit/util/flightcontrol.run)-fm()
      /home/suda/gopath/src/github.com/moby/buildkit/util/flightcontrol/flightcontrol.go:121 +0x41
  sync.(*Once).Do()
      /usr/local/go/src/sync/once.go:44 +0xe1

Previous write at 0x00c4204c6c60 by goroutine 148:
  [failed to restore the stack]

Goroutine 376 (running) created at:
  github.com/moby/buildkit/util/flightcontrol.(*call).wait()
      /home/suda/gopath/src/github.com/moby/buildkit/util/flightcontrol/flightcontrol.go:121 +0x29d
  github.com/moby/buildkit/util/flightcontrol.(*Group).Do()
      /home/suda/gopath/src/github.com/moby/buildkit/util/flightcontrol/flightcontrol.go:56 +0x2b0
  github.com/moby/buildkit/solver.(*sharedOp).Exec()
      /home/suda/gopath/src/github.com/moby/buildkit/solver/jobs.go:651 +0x1bf
  github.com/moby/buildkit/solver.(*edge).execOp()
      /home/suda/gopath/src/github.com/moby/buildkit/solver/edge.go:827 +0x12e
  github.com/moby/buildkit/solver.(*edge).(github.com/moby/buildkit/solver.execOp)-fm()
      /home/suda/gopath/src/github.com/moby/buildkit/solver/edge.go:793 +0x59
  github.com/moby/buildkit/solver/internal/pipe.NewWithFunction.func2()
      /home/suda/gopath/src/github.com/moby/buildkit/solver/internal/pipe/pipe.go:78 +0x5b
==================
--- FAIL: TestSingleCancelParallel (0.98s)
        testing.go:730: race detected during execution of test
--- FAIL: TestHugeGraph (9.74s)
        testing.go:730: race detected during execution of test
```

bk: 2759005f6ad9d3689d30d3ef0f473b75bc72680b
go: 1.10.2

--- a/solver/jobs.go
+++ b/solver/jobs.go
@@ -444,6 +444,7 @@ func (j *Job) Discard() error {
 	j.pw.Close()
 
 	for k, st := range j.list.actives {
+		st.mu.Lock()
 		if _, ok := st.jobs[j]; ok {
 			delete(st.jobs, j)
 			j.list.deleteIfUnreferenced(k, st)
@@ -451,6 +452,7 @@ func (j *Job) Discard() error {
 		if _, ok := st.allPw[j.pw]; ok {
 			delete(st.allPw, j.pw)
 		}
+		st.mu.Unlock()
 	}
 	return nil
 }

